Source/CarthageKit/Project.swift

```diff
@@ -581,16 +583,41 @@ public final class Project {
 					submodule = Submodule(name: project.relativePath, path: project.relativePath, URL: repositoryURLForProject(project, preferHTTPS: self.preferHTTPS), SHA: revision)
 				}
 				
-				if let submodule = submodule {
-					return addSubmoduleToRepository(self.directoryURL, submodule, GitURL(repositoryURL.path!))
-						.startOnQueue(self.gitOperationQueue)
-				} else {
-					return checkoutRepositoryToDirectory(repositoryURL, workingDirectoryURL, revision: revision)
-						.then(self.dependenciesForDependency(dependency))
-						.flatMap(.Merge) { dependencies in
-							return self.symlinkCheckoutPathsForDependencyProject(dependency.project, subDependencies: dependencies, rootDirectoryURL: self.directoryURL)
-						}
+				/// Filter out submodules in `CarthageProjectCheckoutsPath` which would conflict with dependencies.
+				func filter(dependencies: Set<ProjectIdentifier>, by submodules: [Submodule]) -> [ProjectIdentifier] {
+					return dependencies.filter { identifier in
+						!(submodules.filter { (submodule: Submodule) -> Bool in
+							submodule.path.characters.prefix(9).last == .Some("/") && String(
+								submodule.path.characters.prefix(CarthageProjectCheckoutsPath.characters.count)
+							).caseInsensitiveCompare(CarthageProjectCheckoutsPath) == .OrderedSame
```

Comment:
We rely pretty heavily on `NSString.pathComponents` pretty heavily elsewhere, so I think it's fine to do so here as well.